### ISSUE TITLE
Remove wp_head Analytics and AdSense tag verification

### ISSUE THREAD

**Issue by lucymtc** [google/site-kit-wp#83]
## Feature Description

Currently the plugin verifies the existence of Analytics and AdSense tags by first running wp_head hook and parsing the output, if tag was not found then it parses the front end html content, currently being loaded in iframe, to be replaced through fetch request in this PR.

While making the request to the front end site URL wp_head will run either way so the first verification ads complexity to the code and process when is not really needed. 

## Proposed solution

Remove wp_head tag verifications for Analytics https://github.com/google/site-kit-wp/blob/develop/includes/Modules/Analytics.php#L1033-L1044 and AdSense https://github.com/google/site-kit-wp/blob/develop/includes/Modules/AdSense.php#L542-L556 to make the code and process simpler.

## Additional Context

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Lookup for potentially existing Analytics and AdSense tags is never done via manually triggering `wp_head`, `wp_body_open`, or `wp_footer` hook in the backend.
* During Analytics setup, with an existing Analytics tag already added to the frontend via a `wp_head` action, the plugin detects this existing tag correctly and informs the user about it.
* During Analytics setup, with an existing Analytics tag already added to the frontend hard-coded in a random location in the theme template for the front page/home page, the plugin detects this existing tag correctly and informs the user about it.
* During AdSense setup, with an existing AdSense tag already added to the frontend via a `wp_head` action, the plugin detects this existing tag correctly and informs the user about it.
* During AdSense setup, with an existing AdSense tag already added to the frontend hard-coded in a random location in the theme template for the front page/home page, the plugin detects this existing tag correctly and informs the user about it.
* If the user doesn't have access to the Analytics or AdSense property, they are blocked from proceeding.
* If the user does have access to the Analytics or AdSense property, they can't select the property, instead it will be forced to the property identified by the existing tag. The correct account and property are preselected.

## Implementation brief

Existing tag check will be left entirely to the request of frontend which returns all the html content in the response https://github.com/google/site-kit-wp/issues/51
While requesting the frontend any existing tag hardcoded or included through wp_head hook will be detected.

**Analytics**:
- Remove existing tag verification from `get-accounts` datapoint in Analytics https://github.com/google/site-kit-wp/blob/eeec310f16421555f3c9ce8b46daed2e3858b049/includes/Modules/Analytics.php#L1051-L1061
- Remove `tag` datapoint from Analytics that will no longer be used.
- From `tag-permission` datapoint return account id and property instead of `true`, this is data already returned by `has_access_to_property()` so we only need to return the value of `$has_access_to_property` here https://github.com/google/site-kit-wp/blob/eeec310f16421555f3c9ce8b46daed2e3858b049/includes/Modules/Analytics.php#L591
- In analytics/setup.js remove call to `getExistingTag( ‘analytics’ )` inside `getAccounts()` and call in `componentDidMount()`, make request to `tag-permission` all before calling `getAccounts()`. This will save existing tag found and tag-permission response in `sessionStorage`.
- Remove any reference of `responseData.existingTag` inside `getAccounts()` and just set `state.existingTag` with response from `getExistingTag( ‘analytics’ )`.
- `data.deleteCache( 'analytics', 'existingTag' );` should run only run if request to `tag-permission` throws error.
- Add argument to `getAccounts()` with the response of `tag-permission` like `getAccounts( existingTagData )`, where `existingTagData` will contain the account ID and property id, then replace the current checks of `resonseData.existingTag` with `existingTagData` https://github.com/google/site-kit-wp/blob/eeec310f16421555f3c9ce8b46daed2e3858b049/assets/js/modules/analytics/setup.js#L265
- Add optional queryArgs to `get-accounts` request to send the account ID and property ID that we already know from `tag-permission` response.
- Set temporary property `$_existing_tag_account` to the analytics class to store the account id of the existing tag when requesting accounts https://github.com/google/site-kit-wp/blob/eeec310f16421555f3c9ce8b46daed2e3858b049/includes/Modules/Analytics.php#L511 ( will be later removed with https://github.com/google/site-kit-wp/issues/341 )
- Set `$found_account_id` here https://github.com/google/site-kit-wp/blob/develop/includes/Modules/Analytics.php#L1047 with value of `$_existing_tag_account` and remove https://github.com/google/site-kit-wp/blob/eeec310f16421555f3c9ce8b46daed2e3858b049/includes/Modules/Analytics.php#L1051-L1061 
- Ensure this way that account and property are preselected and dropdowns are disabled. 

**AdSense**:
- Remove `tag` datapoint from AdSense that will no longer be used.

AdSense already requests currently the existing tag from the frontend fetch response https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/adsense/dashboard/adsense-module-status.js#L52 and there is no existing request to 'tag' datapoint, so nothing else to remove.
Existing tag is passed as argument to `getAdSenseAccountStatus()` https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/adsense/dashboard/adsense-module-status.js#L57
where user permission checks are done in this condition https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/adsense/util.js#L294 

https://github.com/google/site-kit-wp/issues/89 depends on this issue as the implementation brief has been updated according to the changes of this work removing existing tag verification from the server side.

## Changelog entry

* Simplify and improve accuracy of existing tag detection mechanism by only relying on a remote request.

**Comment by bjackson27** [comment]
@ThierryA @marrrmarrr adding this to the definition board with the note that there is an existing PR for this ticket.

**Comment by felixarntz** [comment]
@bjackson27 Which PR are you referring to?

**Comment by lucymtc** [comment]
@felixarntz https://github.com/google/site-kit-wp/pull/311

**Comment by felixarntz** [comment]
@lucymtc Can you please prioritize writing an implementation brief and working on this? It will affect (and simplify!) both #51 and #110, so I think this should be done first.

That would likely mean though we need to move #51 back to Implementation Brief, as it would be different I assume.

**Comment by felixarntz** [comment]
@lucymtc Implementation brief looks good!

So that means this issue depends on #110 being completed first, right? And then this issue should be completed before #89 I assume (which will need an implementation brief update)?

Would be great if you could clarify your thoughts on the interdependencies. Also make sure to coordinate with the team. I think it's best to reserve all three issues for working on it yourself, since you have the best understanding of how they relate and depend on each other.

**Comment by lucymtc** [comment]
Yes, dependant on https://github.com/google/site-kit-wp/issues/110 for the acceptance criteria:

> If the user doesn't have access to the Analytics or AdSense property, they are blocked from proceeding.
> If the user does have access to the Analytics or AdSense property, they can't select the property, instead it will be forced to the property identified by the existing tag.

https://github.com/google/site-kit-wp/issues/89 is an existing bug but and not directly dependent of this issue, although wanted to mention for when QA is done as it might rise for preselecting property of the existing tag.

**Comment by lilybonney** [comment]
@lucymtc - Could we add a label with what type of issue this is (i.e. bug, enhancement etc.)? Thanks!

**Comment by lucymtc** [comment]
Discussed with @felixarntz and agreed on moving the ticket back to Implementation Brief for next sprint as some requirements have changed and need to be updated and re estimated

**Comment by felixarntz** [comment]
See https://github.com/google/site-kit-wp/pull/311#pullrequestreview-267679938 for additional requirements determined during work on the PR. In order to get the correct properties included in the `get-accounts` response, that datapoint has to be aware of an eventual existing tag (and the property associated with it). Therefore the existing tag detection must happen first.

**Comment by felixarntz** [comment]
@lucymtc Please also have a look at #341 and whether the separation for these two pieces makes sense as it is. They are closely related, and we shouldn't do too much work here that #341 would invalidate.

**Comment by lucymtc** [comment]
thanks @felixarntz 
leaving here as a note, we also need to have a way for the user to refresh in the case they remove the existing tag, right now if you have existing tag and have access to it, even if you remove it, it’s still stored in session storage, so user still gets warned about it

**Comment by lucymtc** [comment]
@felixarntz also assigning this ticket to you so it's in your radar for reviewing the doc with suggested updates for IB.

**Comment by lucymtc** [comment]
@felixarntz assigning to you for Implementation brief review.

**Comment by lucymtc** [comment]
@felixarntz Implementation brief will likely need to be updated after https://github.com/google/site-kit-wp/issues/154 code review is approved ( PR https://github.com/google/site-kit-wp/pull/418 ) moving back to Implementation Brief column

**Comment by lucymtc** [comment]
@felixarntz we need to change the order of steps, currently `getAccounts()` is first called and then `getExistingTag()`  https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics/setup.js#L87
It will likely be fixed with https://github.com/google/site-kit-wp/issues/83

**Comment by lucymtc** [comment]
@felixarntz I have updated Implementation Brief according to the latest changes from https://github.com/google/site-kit-wp/issues/154 assigning back to you for review

**Comment by felixarntz** [comment]
@lucymtc Implementation brief looks great!

> Set temporary property `$_existing_tag_account` to the analytics class to store the account id of the existing tag when requesting accounts

Just something to keep in mind, this is a necessary evil until we fix the architecture via #340

**Comment by felixarntz** [comment]
@lucymtc @aaemnnosttv I just noticed an additional requirement to this one which is highly crucial:

The checks we do on the server-side for an existing tag are more comprehensive than currently on the client side (e.g. they also cover AMP). We need to make sure the JS code detects in the same way, and the existing e2e tests should be expanded to make sure that works.

**Comment by lucymtc** [comment]
@felixarntz I have added the amp tag detection. 
There are different variations of analytics and we only cover 1 of them in e2e. 
Should we expand the test for all variations? if so, should we create a new issue for expanding tag verification e2e tests?

**Comment by felixarntz** [comment]
@lucymtc Good catch! I think this would be better suited for QUnit though. Can you open an issue to add such tests for the `extractTag()` function?

**Comment by felixarntz** [comment]
@tofumatt How did you insert the tag? What was the exact code snippet / markup?

**Comment by felixarntz** [comment]
It seems that this issue messed up the UI for users that don't have a GA account - they still see the button to create a new account, but not the message they are supposed to see.

Here's what it looks like now:
![ga-error](https://user-images.githubusercontent.com/3531426/63863443-8c7cea80-c9ae-11e9-89f6-9966f714d217.png)

Probably the text is missing because no `errorCode` gets set, so the `AnalyticsSetup::renderErrorOrNotice()` function just returns `null`. We can probably just remove that check.

**Comment by felixarntz** [comment]
QA ✅

**Comment by felixarntz** [comment]
Reopening due to minor regression caused by #471, to be fixed by #473
